src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs

```diff
@@ -318,57 +324,129 @@ public static async Task<PackageSpec> GetProjectSpec(IDependencyGraphProject pro
+            var legacyPackageReferenceProjects = new List<IDependencyGraphProject>();
+            var nonLegacyPackageReferenceProjects = new List<IDependencyGraphProject>();
 
-            for (var i = 0; i < projects.Count; i++)
+            foreach (var project in projects)
             {
-                var (packageSpecs, projectAdditionalMessages) = await projects[i].GetPackageSpecsAndAdditionalMessagesAsync(context);
+                if (project.IsLegacyPackageReferenceProject)
+                {
+                    legacyPackageReferenceProjects.Add(project);
+                }
+                else
+                {
+                    await GetProjectRestoreSpecAndAdditionalMessages(new ProjectRestoreSpec(
+                                                project,
+                                                dgSpec,
+                                                context,
+                                                knownProjects,
+                                                allAdditionalMessages));
+                }
+            }
 
-                if (projectAdditionalMessages != null && projectAdditionalMessages.Count > 0)
+            if (legacyPackageReferenceProjects.Count > 0)
+            {
+                var options = new ExecutionDataflowBlockOptions()
                 {
-                    if (allAdditionalMessages == null)
-                    {
-                        allAdditionalMessages = new List<IAssetsLogMessage>();
-                    }
+                    MaxDegreeOfParallelism = Environment.ProcessorCount
+                };
+                var actionBlock = new ActionBlock<ProjectRestoreSpec>(GetProjectRestoreSpecAndAdditionalMessages, options);
 
-                    allAdditionalMessages.AddRange(projectAdditionalMessages);
+                foreach (IDependencyGraphProject project in legacyPackageReferenceProjects)
+                {
+                    await actionBlock.SendAsync(new ProjectRestoreSpec(
+                                                    project,
+                                                    dgSpec,
+                                                    context,
+                                                    knownProjects,
+                                                    allAdditionalMessages)
+                                                );
                 }
 
-                foreach (var packageSpec in packageSpecs)
+                actionBlock.Complete();
+                await actionBlock.Completion;
+            }
+
+            // Return dg file
+            return (dgSpec, allAdditionalMessages.ToList());
+        }
+
+        private async static Task GetProjectRestoreSpecAndAdditionalMessages(
+            ProjectRestoreSpec restoreSpecData)
+        {
+            IReadOnlyList<PackageSpec> packageSpecs;
+            IReadOnlyList<IAssetsLogMessage> projectAdditionalMessages;
+
+            await SemaphoreLock.WaitAsync();
```

Comment:
If we are locking here, are we really paralellizing?

The expensive part of legacy package reference get package specs is the call in question. 
if we are synchronizing on that we're never really running the expensive part in parallel, are we? 

Am I missing something.

Reply:
Problem here is we're passing around `restoreSpecData.Context` as parameter, even though it's cache context somehow it's [not using Concurrent dictionary](https://github.com/NuGet/NuGet.Client/blob/9ed028c88d0e6437b9747bb3983302a3f82ac518/src/NuGet.Core/NuGet.PackageManagement/DependencyGraphCacheContext.cs#L35) which safe for concurrency. 
https://github.com/NuGet/NuGet.Client/blob/9ed028c88d0e6437b9747bb3983302a3f82ac518/src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs#L389

I already have [issue#10121](https://github.com/NuGet/Home/issues/10121) for this problem.

Above call is doing mutation inside that method here:
https://github.com/NuGet/NuGet.Client/blob/9ed028c88d0e6437b9747bb3983302a3f82ac518/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/CpsPackageReferenceProject.cs#L188-L190

If multiple threads try to insert new values same time they can break bucket for keys(linkedlist) in dictionary, therefore we'll get broken bucket(linked listed with self loop) by rare random chance. Theoretically it's very rare but did happen to me previously, system hang up every 2-3 months, and someone have to intervene. Once bucket breaks it causes infinite loop when you try to get value from it. Our solution was `Concurrent dictionary` instead of normal dictionary. 
Since we can't change already public api solution would be this `SemaphoreLock` lock. I tried to use normal `lock` it doesn't work nicely with async call, therefore had to use `SemaphoreLock` lock instead. Actually this is my first time using this `SemaphoreLock`, if it's improper usage then please correct me. 🙏 

**_Actually I did check with check with and without this SemaphoreLocklock there was almost no difference, we're still getting almost same amount of perf gain._**

Reply:
I am doing additional measurements on my laptop now which is very slow. Let's see what happens on it.

Reply:
Did you consider locking on the cache context *only* in the legacy packagereference projects?

What are the numbers then?

I want us to focus on parallelizing the correct things. If our synchronization is too coarse, we might not be getting any improvements, yet we'd be spamming the threadpool.

Reply:
At least RPS test passes, so we're not making things worse. 😄 
https://dev.azure.com/DevDiv/DevDiv/_releaseProgress?releaseId=855305&_a=release-pipeline-progress

Reply:
I don't think the RPS test is sensitive to these changes. 

We also have no test cases for legacy projects yet.

Reply:
@nkolev92 here are measurements from **laptop** which is quite slow.
Parallelization only works for legacy packagereference projects and it looks parallelization makes things consistently worse in all cases with and without `Semaphore lock`.
It looks we need to look other ways besides parallelization.

4 Legacy | Sequential | Non-Semaphore | Semaphore
-- | -- | -- | --
Avg | 119.65 | 146.925 | 214.475
StDev | 122.208 | 153.6335 | 188.8585


21 Legacy project | Sequential | Non-Semaphore | Semaphore
-- | -- | -- | --
Avg | 337.68 | 538.88 | 454.9
StDev | 259.264379 | 283.6895 | 265.0791


43 Mixed (legacy+package.config) | Sequential | Non-Semaphore | Semaphore
-- | -- | -- | --
Avg | 513.9 | 546.22 | 571.26
StDev | 299.7865 | 284.7266 | 287.5068


Here are measurements from **desktop** which is very fast, are some improvements but still not consistent.
4 Legacy project | Sequential | Non-Semaphore | Semaphore
-- | -- | -- | --
Avg | 69.075 | 48.075 | 77.775
StDev | 59.62377 | 35.97391 | 77.19887

21 Legacy project | Sequential | Non-Semaphore | Semaphore
-- | -- | -- | --
Avg | 468.82 | 501.6 | 400.02
StDev | 214.1120906 | 237.0053818 | 185.171789


43 Mixed (legacy+package.config) | Sequential | Non-Semaphore | Semaphore
-- | -- | -- | --
Avg | 544 | 540.26 | 514.4
StDev | 289.1091 | 325.8691 | 326.7171

Reply:
@nkolev92 
Here root of all slowness is GetPackageSpecAsync method. It's calling many IVsAdapters/IVsBuildPropertyStorage which are very slow, not sure if they really work in parallel actually. I couldn't find source file for them. It's not in our  repo.

Reply:
Thanks for all this hard work @erdembayar!

Let's have a quick chat tomorrow after 3pm, but at this point, the noise in the results to me suggests that parallelization might not be worth it right now, and it's likely not the answer to our problems, rather improving the GetPackageSpecs method for Legacy PackageReference projects in general.

What do you think?

Reply:
@nkolev92 
Sure, I'll ping you around 3.00pm

Reply:
@nkolev92  @aortiz-msft 
Nikolche and I had discussion on this. We're pretty much done here. So moving on to next question "can we do something to improve getpackagespecs for legacy package reference project?"
We need more measurements and there is no low hanging fruit therefore we create another new issue[#10197](https://github.com/NuGet/Home/issues/10197)  for that task and decided to conclude/close this PR/issue now.